File: examples/algorithms/montecarlo_uniform.py

#title: Estimate mean with given confidence interval range using Monte Carlo
#author: Yann Richet
#type: sampling
#options: batch_sample_size=10;max_iterations=100;confidence=0.9;target_confidence_range=1.0;seed=42
#require: numpy;scipy;matplotlib

class MonteCarlo_Uniform:
    """Monte Carlo sampling algorithm with adaptive stopping based on confidence interval"""

    def __init__(self, **options):
        """Initialize with algorithm options"""
        self.options = {}
        self.options["batch_sample_size"] = int(options.get("batch_sample_size", 10))
        self.options["max_iterations"] = int(options.get("max_iterations", 100))
        self.options["confidence"] = float(options.get("confidence", 0.9))
        self.options["target_confidence_range"] = float(options.get("target_confidence_range", 1.0))

        self.n_samples = 0
        self.variables = {}

        import numpy as np
        np.random.seed(int(options.get("seed", 42)))

    def get_initial_design(self, input_variables, output_variables):
        """
        Generate initial design

        Args:
            input_variables: Dict[str, Tuple[float, float]] - {var: (min, max)}
            output_variables: List[str] - output variable names
        """
        for v, bounds in input_variables.items():
            # Bounds are already parsed as tuples (min, max)
            if isinstance(bounds, tuple) and len(bounds) == 2:
                self.variables[v] = bounds
            else:
                raise ValueError(
                    f"Input variable {v} must have (min, max) tuple bounds for MonteCarlo_Uniform sampling"
                )
        return self._generate_samples(self.options["batch_sample_size"])
  
    def get_next_design(self, X, Y):
        """
        Generate next design based on convergence criteria

        Args:
            X: List[Dict[str, float]] - previous inputs
            Y: List[float] - previous outputs (may contain None)

        Returns:
            List[Dict[str, float]] - next points, or [] if finished
        """
        # Check max iterations
        if self.n_samples >= self.options["max_iterations"] * self.options["batch_sample_size"]:
            return []

        # Filter out None values
        import numpy as np
        from scipy import stats
        Y_valid = [y for y in Y if y is not None]

        if len(Y_valid) < 2:
            return self._generate_samples(self.options["batch_sample_size"])

        Y_array = np.array(Y_valid)
        mean = np.mean(Y_array)
        conf_int = stats.t.interval(
            self.options["confidence"],
            len(Y_array) - 1,
            loc=mean,
            scale=stats.sem(Y_array)
        )
        conf_range = conf_int[1] - conf_int[0]

        # Stop if confidence interval is narrow enough
        if conf_range <= self.options["target_confidence_range"]:
            return []

        # Generate more samples
        return self._generate_samples(self.options["batch_sample_size"])
      
    def _generate_samples(self, n):
        import numpy as np
        samples = []
        for _ in range(n):
            sample = {}
            for v,(min_val,max_val) in self.variables.items():
                sample[v] = np.random.uniform(min_val, max_val)
            samples.append(sample)
        self.n_samples += n
        return samples
  
    def get_analysis(self, X, Y):
        """
        Display results with statistics and histogram

        Args:
            X: List[Dict[str, float]] - all evaluated inputs
            Y: List[float] - all outputs (may contain None)

        Returns:
            Dict with 'text', 'data', and optionally 'html' keys
        """
        import numpy as np
        from scipy import stats

        display_dict = {"text": "", "data": {}}

        # Filter out None values
        Y_valid = [y for y in Y if y is not None]

        if len(Y_valid) < 2:
            display_dict["text"] = "Not enough valid results to display statistics"
            display_dict["data"] = {"valid_samples": len(Y_valid)}
            return display_dict

        Y_array = np.array(Y_valid)
        mean = np.mean(Y_array)
        std = np.std(Y_array)
        conf_int = stats.t.interval(
            self.options["confidence"],
            len(Y_array) - 1,
            loc=mean,
            scale=stats.sem(Y_array)
        )

        # Store data
        display_dict["data"] = {
            "mean": float(mean),
            "std": float(std),
            "confidence_interval": [float(conf_int[0]), float(conf_int[1])],
            "n_samples": len(Y_valid),
            "min": float(np.min(Y_array)),
            "max": float(np.max(Y_array))
        }

        # Create text summary
        display_dict["text"] = f"""Monte Carlo Sampling Results:
  Valid samples: {len(Y_valid)}
  Mean: {mean:.6f}
  Std: {std:.6f}
  {self.options['confidence']*100:.0f}% confidence interval: [{conf_int[0]:.6f}, {conf_int[1]:.6f}]
  Range: [{np.min(Y_array):.6f}, {np.max(Y_array):.6f}]
"""

        # Try to create HTML with histogram
        try:
            import matplotlib
            matplotlib.use('Agg')  # Non-interactive backend
            import matplotlib.pyplot as plt
            import base64
            from io import BytesIO

            plt.figure(figsize=(8, 6))
            plt.hist(Y_array, bins=20, density=True, alpha=0.6, color='g', edgecolor='black')
            plt.title("Output Distribution")
            plt.xlabel("Output Value")
            plt.ylabel("Density")
            plt.grid(alpha=0.3)

            # Add mean line
            plt.axvline(mean, color='r', linestyle='--', linewidth=2, label=f'Mean: {mean:.3f}')
            plt.legend()

            # Convert to base64
            buffered = BytesIO()
            plt.savefig(buffered, format="png", dpi=100, bbox_inches='tight')
            plt.close()
            img_str = base64.b64encode(buffered.getvalue()).decode()

            html_output = f"""<div>
  <p><strong>Estimated mean:</strong> {mean:.6f}</p>
  <p><strong>{self.options['confidence']*100:.0f}% confidence interval:</strong> [{conf_int[0]:.6f}, {conf_int[1]:.6f}]</p>
  <img src="data:image/png;base64,{img_str}" alt="Histogram" style="max-width:800px;"/>
</div>"""
            display_dict["html"] = html_output
        except Exception as e:
            # If plotting fails, just skip it
            pass

        return display_dict

    def get_analysis_tmp(self, X, Y):
        """
        Display intermediate results at each iteration

        Args:
            X: List[Dict[str, float]] - all evaluated inputs so far
            Y: List[float] - all outputs so far (may contain None)

        Returns:
            Dict with 'text' and 'data' keys
        """
        import numpy as np
        from scipy import stats

        # Filter out None values
        Y_valid = [y for y in Y if y is not None]

        if len(Y_valid) < 2:
            return {
                'text': f"  Progress: {len(Y_valid)} valid sample(s) collected",
                'data': {'valid_samples': len(Y_valid)}
            }

        Y_array = np.array(Y_valid)
        mean = np.mean(Y_array)
        std = np.std(Y_array)
        conf_int = stats.t.interval(
            self.options["confidence"],
            len(Y_array) - 1,
            loc=mean,
            scale=stats.sem(Y_array)
        )
        conf_range = conf_int[1] - conf_int[0]

        return {
            'text': f"  Progress: {len(Y_valid)} samples, "
                   f"mean={mean:.6f}, "
                   f"{self.options['confidence']*100:.0f}% CI range={conf_range:.6f}",
            'data': {
                'n_samples': len(Y_valid),
                'mean': float(mean),
                'std': float(std),
                'confidence_range': float(conf_range)
            }
        }

File: examples/algorithms/montecarlo_uniform.R

#title: Estimate mean with given confidence interval range using Monte Carlo
#author: Yann Richet
#type: sampling
#options: batch_sample_size=10;max_iterations=100;confidence=0.9;target_confidence_range=1.0;seed=42
#require: base64enc

# Constructor for MonteCarlo_Uniform S3 class
MonteCarlo_Uniform <- function(...) {
  # Get options from ... arguments
  opts <- list(...)

  # Create object with initial state
  # Use an environment for mutable state (idiomatic S3 pattern)
  state <- new.env(parent = emptyenv())
  state$n_samples <- 0
  state$variables <- list()

  obj <- list(
    options = list(
      batch_sample_size = as.integer(
        ifelse(is.null(opts$batch_sample_size), 10, opts$batch_sample_size)
      ),
      max_iterations = as.integer(
        ifelse(is.null(opts$max_iterations), 100, opts$max_iterations)
      ),
      confidence = as.numeric(
        ifelse(is.null(opts$confidence), 0.9, opts$confidence)
      ),
      target_confidence_range = as.numeric(
        ifelse(is.null(opts$target_confidence_range), 1.0, opts$target_confidence_range)
      )
    ),
    state = state  # Environment for mutable state
  )

  # Set random seed
  seed <- ifelse(is.null(opts$seed), 42, opts$seed)
  set.seed(as.integer(seed))

  # Set S3 class
  class(obj) <- "MonteCarlo_Uniform"

  return(obj)
}

# Generic function definitions (if not already defined)
if (!exists("get_initial_design")) {
  get_initial_design <- function(obj, ...) UseMethod("get_initial_design")
}

if (!exists("get_next_design")) {
  get_next_design <- function(obj, ...) UseMethod("get_next_design")
}

if (!exists("get_analysis")) {
  get_analysis <- function(obj, ...) UseMethod("get_analysis")
}

if (!exists("get_analysis_tmp")) {
  get_analysis_tmp <- function(obj, ...) UseMethod("get_analysis_tmp")
}

# Method: get_initial_design
get_initial_design.MonteCarlo_Uniform <- function(obj, input_variables, output_variables) {
  # Store variable bounds in mutable state
  # input_variables is a named list: list(var1 = c(min, max), var2 = c(min, max))
  for (v in names(input_variables)) {
    bounds <- input_variables[[v]]
    if (!is.numeric(bounds) || length(bounds) != 2) {
      stop(paste("Input variable", v, "must have c(min, max) bounds for MonteCarlo_Uniform sampling"))
    }
    obj$state$variables[[v]] <- bounds
  }

  return(generate_samples(obj, obj$options$batch_sample_size))
}

# Method: get_next_design
get_next_design.MonteCarlo_Uniform <- function(obj, X, Y) {
  # Check max iterations
  if (obj$state$n_samples >= obj$options$max_iterations * obj$options$batch_sample_size) {
    return(list())  # Empty list signals finished
  }

  # Filter out NULL/NA values
  Y_valid <- Y[!sapply(Y, is.null) & !is.na(Y)]
  Y_valid <- unlist(Y_valid)

  if (length(Y_valid) < 2) {
    return(generate_samples(obj, obj$options$batch_sample_size))
  }

  # Calculate confidence interval
  mean_y <- mean(Y_valid)
  n <- length(Y_valid)
  se <- sd(Y_valid) / sqrt(n)

  # t-distribution confidence interval
  alpha <- 1 - obj$options$confidence
  t_critical <- qt(1 - alpha/2, df = n - 1)
  conf_int_lower <- mean_y - t_critical * se
  conf_int_upper <- mean_y + t_critical * se
  conf_range <- conf_int_upper - conf_int_lower

  # Stop if confidence interval is narrow enough
  if (conf_range <= obj$options$target_confidence_range) {
    return(list())  # Finished
  }

  # Generate more samples
  return(generate_samples(obj, obj$options$batch_sample_size))
}

# Method: get_analysis
get_analysis.MonteCarlo_Uniform <- function(obj, X, Y) {
  analysis_dict <- list(text = "", data = list())

  # Filter out NULL/NA values
  Y_valid <- Y[!sapply(Y, is.null) & !is.na(Y)]
  Y_valid <- unlist(Y_valid)

  if (length(Y_valid) < 2) {
    analysis_dict$text <- "Not enough valid results to analyze statistics"
    analysis_dict$data <- list(valid_samples = length(Y_valid))
    return(analysis_dict)
  }

  # Calculate statistics
  mean_y <- mean(Y_valid)
  std_y <- sd(Y_valid)
  n <- length(Y_valid)
  se <- std_y / sqrt(n)

  # t-distribution confidence interval
  alpha <- 1 - obj$options$confidence
  t_critical <- qt(1 - alpha/2, df = n - 1)
  conf_int_lower <- mean_y - t_critical * se
  conf_int_upper <- mean_y + t_critical * se

  # Store data
  analysis_dict$data <- list(
    mean = mean_y,
    std = std_y,
    confidence_interval = c(conf_int_lower, conf_int_upper),
    n_samples = length(Y_valid),
    min = min(Y_valid),
    max = max(Y_valid)
  )

  # Create text summary
  analysis_dict$text <- sprintf(
"Monte Carlo Sampling Results:
  Valid samples: %d
  Mean: %.6f
  Std: %.6f
  %.0f%% confidence interval: [%.6f, %.6f]
  Range: [%.6f, %.6f]
",
    length(Y_valid),
    mean_y,
    std_y,
    obj$options$confidence * 100,
    conf_int_lower,
    conf_int_upper,
    min(Y_valid),
    max(Y_valid)
  )

  # Try to create HTML with histogram
  tryCatch({
    # Create histogram plot
    png_file <- tempfile(fileext = ".png")
    png(png_file, width = 800, height = 600)

    hist(Y_valid, breaks = 20, freq = FALSE,
         col = rgb(0, 1, 0, 0.6),
         border = "black",
         main = "Output Distribution",
         xlab = "Output Value",
         ylab = "Density")
    grid(col = rgb(0, 0, 0, 0.3))

    # Add mean line
    abline(v = mean_y, col = "red", lwd = 2, lty = 2)
    legend("topright",
           legend = sprintf("Mean: %.3f", mean_y),
           col = "red", lty = 2, lwd = 2)

    dev.off()

    # Convert to base64
    if (requireNamespace("base64enc", quietly = TRUE)) {
      img_base64 <- base64enc::base64encode(png_file)

      html_output <- sprintf(
'<div>
  <p><strong>Estimated mean:</strong> %.6f</p>
  <p><strong>%.0f%% confidence interval:</strong> [%.6f, %.6f]</p>
  <img src="data:image/png;base64,%s" alt="Histogram" style="max-width:800px;"/>
</div>',
        mean_y,
        obj$options$confidence * 100,
        conf_int_lower,
        conf_int_upper,
        img_base64
      )
      analysis_dict$html <- html_output
    }

    # Clean up temp file
    unlink(png_file)
  }, error = function(e) {
    # If plotting fails, just skip it
  })

  return(analysis_dict)
}

# Method: get_analysis_tmp
get_analysis_tmp.MonteCarlo_Uniform <- function(obj, X, Y) {
  # Filter out NULL/NA values
  Y_valid <- Y[!sapply(Y, is.null) & !is.na(Y)]
  Y_valid <- unlist(Y_valid)

  if (length(Y_valid) < 2) {
    return(list(
      text = sprintf("  Progress: %d valid sample(s) collected", length(Y_valid)),
      data = list(valid_samples = length(Y_valid))
    ))
  }

  # Calculate statistics
  mean_y <- mean(Y_valid)
  std_y <- sd(Y_valid)
  n <- length(Y_valid)
  se <- std_y / sqrt(n)

  # t-distribution confidence interval
  alpha <- 1 - obj$options$confidence
  t_critical <- qt(1 - alpha/2, df = n - 1)
  conf_int_lower <- mean_y - t_critical * se
  conf_int_upper <- mean_y + t_critical * se
  conf_range <- conf_int_upper - conf_int_lower

  return(list(
    text = sprintf(
      "  Progress: %d samples, mean=%.6f, %.0f%% CI range=%.6f",
      length(Y_valid),
      mean_y,
      obj$options$confidence * 100,
      conf_range
    ),
    data = list(
      n_samples = length(Y_valid),
      mean = mean_y,
      std = std_y,
      confidence_range = conf_range
    )
  ))
}

# Helper function: generate_samples (not a method, internal use only)
generate_samples <- function(obj, n) {
  samples <- list()

  for (i in 1:n) {
    sample <- list()
    for (v in names(obj$state$variables)) {
      bounds <- obj$state$variables[[v]]
      sample[[v]] <- runif(1, min = bounds[1], max = bounds[2])
    }
    samples[[i]] <- sample
  }

  # Update n_samples in state environment (mutable)
  obj$state$n_samples <- obj$state$n_samples + n

  return(samples)
}
